# KEGGDecoder Binder Example

This repository generates acalculate metabolic pathway completeness and build heatmaps from [GhostKOALA](https://www.kegg.jp/ghostkoala/) output via KEGGDecoder.

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/taylorreiter/KEGGDecoder-binder/master)

This binder provides instructions and demo for the [KEGGDecoder](https://github.com/bjtully/BioData/tree/master/KEGGDecoder) tool, which makes a metabolic heatmap. The package was developed by Ben Tully and expanded during the [Speeding Up Science workshop](https://speeding-up-science-workshops.github.io/speeding-up-science/) by Jay Osvatic, Roth Conrad, Luiz Irber, Taylor Reiter, Chris Neely, Jason Fell, and Marisa Lim. 

## Quick Start
1) Launch the binder with the button above. This will take a few minutes to load.
2) Upload your ghostKOALA or kofamscan output to the Jupyter Notebook in your binder session with the `Upload` button near the top right of that page.
3) Select `New Python 3 Notebook` from the `New` button next to the `Upload` button. This will open in a new tab.
4) From this Jupyter notebook, execute:  
`!KEGG-decoder --input Example_input_data.txt --output Example_output_data.tsv --vizoption interactive`
5) Return to the first binder session tab and click on the new `function_heatmap.html` file to open it in a new tab.
6) Explore.

For instructions on how to generate the input data needed for KEGGDecoder, see the notebooks in example_notebooks 
directory.



KEGGDecoder - Visualizing biogeochemical metabolic pathways
====
filename: README.md

## Quick Start

  - Any instructions for running. e.g. does user need to open the binder and then click "knit"?
  - 
  -
## Example Output

   - ![Example Output](Example_Output_Image.png)

## Summary

   - Designed to parse and visualize a KEGG-Koala output to determine the completeness of biogeochemically-relevant metabolic pathways
   	- metabolic pathways are hand-curated to select for informative markers of predicted function
   - The immediate input for KEGGDecoder is a tab-delimited KEGG-Koala output file that consists of a formatted gene ID and a correspoding KEGG Ontology assignment
   	- Options for generating a KEGG-Koala output include:
   		- [BlastKOALA](https://www.kegg.jp/blastkoala/)
   		- [GhostKOALA](https://www.kegg.jp/ghostkoala/)
   		- [KofamScan](https://www.genome.jp/tools/kofamkoala/)
   	- KEGG Ontology assignment requires a protein fasta file for submission and can be generated by tools such as:
   		- [Prodigal](https://dx.doi.org/10.1186%2F1471-2105-11-119)
   		- [MetaSanity](https://github.com/cjneely10/MetaSanity) - which can generate both a protein FASTA file and a KEGG-Koala output for a set of genomes
   - 
 
 
## Authors

 - Author Name, github id (and ORCID if you have it)
 - Benjamin Tully, bjtully, 0000-0002-9384-7635
 - Taylor Reiter, taylorreiter, 0000-0002-7388-421X
 - Luiz Irber, luizirber, 0000-0003-4371-9659
 - Roth Conrad, rotheconrad, 0000-0001-8155-8441 
 - Jay Osvatic, osvatic, 0000-0002-7765-0058
 - Chris Neely, cjnelly10, 0000-0002-2620-8948
 - Marisa Lim,
 - Jason Fell, jfell13, 0000-0001-6680-2936


## Links

Zenodo Binder, doi: LINK_TO_BINDER

Github Binder: LINK_TO_BINDER

Github Repository: https://github.com/speeding-up-science-workshops/<REPO_NAME>


